scrape/manager_test.go

```diff
@@ -714,3 +725,164 @@ scrape_configs:
 	reload(scrapeManager, cfg2)
 	require.ElementsMatch(t, []string{"job1", "job3"}, scrapeManager.ScrapePools())
 }
+
+func TestManagerScrapeCreatedTimestamp(t *testing.T) {
+	counterType := dto.MetricType_COUNTER
+	now := time.Now()
+	nowMs := now.UnixMilli()
+	testCases := []struct {
+		name     string
+		mf       *dto.MetricFamily
+		ingestCT bool
+		mfName   string
+		outputs  []float64
+	}{
+		{
+			name:   "valid counter - flag true",
+			mfName: "expected_counter",
+			mf: &dto.MetricFamily{
+				Name: proto.String("expected_counter"),
+				Type: &counterType,
+				Metric: []*dto.Metric{
+					{
+						Counter: &dto.Counter{
+							Value:            proto.Float64(1.0),
+							CreatedTimestamp: timestamppb.New(now),
+						},
+					},
+				},
+			},
+			ingestCT: true,
+			outputs:  []float64{0.0, 1.0},
+		},
+		{
+			name:   "valid counter - flag false",
+			mfName: "expected_counter",
+			mf: &dto.MetricFamily{
+				Name: proto.String("expected_counter"),
+				Type: &counterType,
+				Metric: []*dto.Metric{
+					{
+						Counter: &dto.Counter{
+							Value:            proto.Float64(1.0),
+							CreatedTimestamp: timestamppb.New(now),
+						},
+					},
+				},
+			},
+			ingestCT: false,
+			outputs:  []float64{1.0},
+		},
+		{
+			name:   "created timestamp older than sample timestamp",
+			mfName: "expected_counter",
+			mf: &dto.MetricFamily{
+				Name: proto.String("expected_counter"),
+				Type: &counterType,
+				Metric: []*dto.Metric{
+					{
+						TimestampMs: &nowMs,
+						Counter: &dto.Counter{
+							Value:            proto.Float64(1.0),
+							CreatedTimestamp: timestamppb.New(now.Add(time.Hour)),
+						},
+					},
+				},
+			},
+			ingestCT: true,
+			outputs:  []float64{1.0},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			app := &collectResultAppender{}
+			reg := prometheus.NewRegistry()
+			scrapeManager, err := NewManager(&Options{EnableCreatedTimestampIngestion: tc.ingestCT}, log.NewLogfmtLogger(os.Stderr), &collectResultAppendable{app}, reg)
+			require.NoError(t, err)
+
+			require.NoError(t, scrapeManager.ApplyConfig(&config.Config{
+				GlobalConfig: config.GlobalConfig{
+					ScrapeInterval:  model.Duration(5 * time.Second),
+					ScrapeTimeout:   model.Duration(5 * time.Second),
+					ScrapeProtocols: []config.ScrapeProtocol{config.PrometheusProto},
+				},
+				ScrapeConfigs: []*config.ScrapeConfig{{JobName: "test"}},
+			}))
+
+			buf := &bytes.Buffer{}
+			protoBuf, err := proto.Marshal(tc.mf)
+
+			// Write first length, then binary protobuf.
```

Comment:
@TheSpiritXIII once showed me an article(or was it stackoverflow?), that I can't seem to find anymore, that shows how protobuf needs its length attached as prefix.

We need to do this since prometheus has its own marshal/unmarshal implementation. If we don't add it, indeed the tests fail during unmarshalling. Also done in protobuf parser tests: 

https://github.com/prometheus/prometheus/blob/6722a5bb8613fe86994241cb8ebe1fdda5d317e3/model/textparse/protobufparse_test.go#L600-L617